data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java

```diff
@@ -339,139 +396,57 @@ public InputStream executeStreamingQuery(String database, String command) throws
-                        jsonPayload, timeoutMs + CLIENT_SERVER_DELTA_IN_MILLISECS, headers),
-                "ClientImpl.executeStreamingQuery", updateAndGetExecuteTracingAttributes(database, properties));
+        KustoRequest kr = new KustoRequest(command, database, properties);
+        return executeStreamingQuery(kr);
     }
 
-    private long determineTimeout(ClientRequestProperties properties, CommandType commandType, String clusterUrl) throws DataClientException {
-        Long timeoutMs;
-        try {
-            timeoutMs = properties == null ? null : properties.getTimeoutInMilliSec();
-        } catch (ParseException e) {
-            throw new DataClientException(clusterUrl, "Failed to parse timeout from ClientRequestProperties");
-        }
-        if (timeoutMs == null) {
-            if (commandType == CommandType.ADMIN_COMMAND) {
-                timeoutMs = COMMAND_TIMEOUT_IN_MILLISECS;
-            } else {
-                timeoutMs = QUERY_TIMEOUT_IN_MILLISECS;
-            }
-        }
-        return timeoutMs;
-    }
-
-    private CommandType determineCommandType(String command) {
-        if (command.startsWith(ADMIN_COMMANDS_PREFIX)) {
-            return CommandType.ADMIN_COMMAND;
-        }
-        return CommandType.QUERY;
-    }
-
-    private Map<String, String> generateIngestAndCommandHeaders(ClientRequestProperties properties,
-            String clientRequestIdPrefix,
-            String activityTypeSuffix)
-            throws DataServiceException, DataClientException {
-
-        Map<String, String> headers = extractTracingHeaders(properties);
-
-        if (aadAuthenticationHelper != null) {
-            headers.put(HttpHeaders.AUTHORIZATION, String.format("Bearer %s", aadAuthenticationHelper.acquireAccessToken()));
-        }
-
-        String clientRequestId;
-        if (properties != null && StringUtils.isNotBlank(properties.getClientRequestId())) {
-            clientRequestId = properties.getClientRequestId();
-        } else {
-            clientRequestId = String.format("%s;%s", clientRequestIdPrefix, UUID.randomUUID());
-        }
-
-        headers.put("x-ms-client-request-id", clientRequestId);
-
-        headers.put("Connection", "Keep-Alive");
-
-        UUID activityId = UUID.randomUUID();
-        String activityContext = String.format("%s%s/%s, ActivityId=%s, ParentId=%s, ClientRequestId=%s", JAVA_INGEST_ACTIVITY_TYPE_PREFIX, activityTypeSuffix,
-                activityId, activityId, activityId, clientRequestId);
-        headers.put("x-ms-activitycontext", activityContext);
+    public InputStream executeStreamingQuery(@NotNull KustoRequest kr) throws DataServiceException, DataClientException {
 
-        // replace non-ascii characters in header values with '?'
-        headers.replaceAll((_i, v) -> v == null ? null : v.replaceAll("[^\\x00-\\x7F]", "?"));
-        return headers;
-    }
+        // Validate and optimize the query object
+        kr.validateAndOptimize();
```

Comment:
We could but then we'd have to set content type explicitly in prepareRequest. In this instance the content type is set explicitly on the builder, in the other method I used the implicit content type which is just JSON. This would save lines though. The other thing is the prepareRequest assumes you don't want to set any special headers, whereas this one sets special headers for streaming content. I probably never should have made a prepareRequest method in the first place but I figured it would be useful for testability. It's not there for reuse so much as its existence is for breaking down what's happening step by step.

Reply:
Maybe renaming that method to prepareJSONRequest would be better and then creating a prepareStreamingRequest method for this logic would make more sense?